Cleanup placeholder replacement

rjernst opened the pull request:
This change moves placeholder replacement to a pkg private class for
settings. It also adds a null check when calling replacement, as
settings objects can still contain null values, because we only prohibit
nulls on file loading. Finally, this cleans up file and stream loading a
bit to not have unnecessary exception wrapping.

rjernst committed:
Settings: Cleanup placeholder replacement

This change moves placeholder replacement to a pkg private class for
settings. It also adds a null check when calling replacement, as
settings objects can still contain null values, because we only prohibit
nulls on file loading. Finally, this cleans up file and stream loading a
bit to not have unnecessary exception wrapping.

jasontedor commented on core/src/main/java/org/elasticsearch/common/logging/LogConfigurator.java:
```diff
@@ -167,7 +167,7 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
         try {
             settingsBuilder.loadFromPath(file);
-        } catch (SettingsException | NoClassDefFoundError e) {
+        } catch (IOException | SettingsException | NoClassDefFoundError e) {
```
Any idea why the `NoClassDefFoundError` is here? Is this leftover and no longer needed after the log4j-only migration?

rjernst replied:
Possibly, not sure..

jasontedor commented on core/src/main/java/org/elasticsearch/common/settings/PropertyPlaceholder.java:
```diff
@@ -34,24 +34,13 @@
  * {@link PlaceholderResolver}.
  */
-public class PropertyPlaceholder {
+class PropertyPlaceholder {
```
Since I can't leave emoji on a line of code without leaving a comment, this comment is so that I can: :+1:

jasontedor commented on core/src/test/java/org/elasticsearch/common/settings/loader/YamlSettingsLoaderTests.java:
```diff
@@ -48,42 +53,39 @@ public void testSimpleYamlSettings() throws Exception {
+        ElasticsearchParseException e = expectThrows(ElasticsearchParseException.class, () -> {
+            settingsBuilder().loadFromPath(tmp);
+        });
+        assertTrue(e.getMessage(), e.getMessage().contains("null-valued setting found for key [foo]"));
```
Can we keep the line numbers in the test assertions? I think they are important to maintain.

rjernst replied:
Ok, I expanded the string a little so it contains "found at line number", but I don't think the particular line number is important, just that we have the line number in the exception message?

rjernst committed:
Merge branch 'master' into placeholder

rjernst committed:
Check more complete exception message for missing setting value

rjernst commented:
@jasontedor I pushed a new commit.

clintongormley commented:
@jasontedor please could you review this again?

jasontedor committed:
Merge branch 'master' into rjernst-placeholder

* master: (911 commits)
  [TEST] wait for yellow after setup doc tests (#18726)
  Fix recovery throttling to properly handle relocating non-primary shards (#18701)
  Fix merge stats rendering in RestIndicesAction (#18720)
  [TEST] mute RandomAllocationDeciderTests.testRandomDecisions
  Reworked docs for index-shrink API (#18705)
  Improve painless compile-time exceptions
  Adds UUIDs to snapshots
  Add test rethrottle test case for delete-by-query
  Do not start scheduled pings until transport start
  Adressing review comments
  Only filter intial recovery (post API) when shrinking an index (#18661)
  Add tests to check that toQuery() doesn't return null
  Removing handling of null lucene query where we catch this at parse time
  Handle empty query bodies at parse time and remove EmptyQueryBuilder
  Mute failing assertions in IndexWithShadowReplicasIT until fix
  Remove allow running as root
  Add upgrade-not-supported warning to alpha release notes
  remove unrecognized javadoc tag from matrix aggregation module
  set ValuesSourceConfig fields as private
  Adding MultiValuesSource support classes and documentation to matrix stats agg module
  ...

jasontedor committed:
Fix uncaught checked exception in AzureTestUtils

This commit fixes an uncaught checked IOException now thrown in
AzureTestUtils after 3adaf096758a6015ca4f733e2e49ee5528ac3cd5.

jasontedor committed:
More complete exception message in settings tests

This commit adds additional details to the exception message assertions
in the YAML settings loader tests.

jasontedor commented:
I merged master in, fixed a test in 974c753bf6fdc00d30bab96d737447a0b9113a8f and added some additional detail to the exception message assertions in be0036542c4d1c40c0044cd22f45ea1d2534b8ec. Feel free to revert this last one if you don't agree. LGTM.
